snapcraft/internal/repo/apt_sources_manager.py

```diff
@@ -0,0 +1,231 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2015-2021 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import io
+import logging
+import os
+import pathlib
+import re
+import subprocess
+import tempfile
+from typing import List, Optional
+
+from snapcraft.internal import os_release
+from snapcraft.internal.meta import package_repository
+from snapcraft.project._project_options import ProjectOptions
+
+from . import apt_ppa
+
+logger = logging.getLogger(__name__)
+
+
+def _sudo_write_file(*, dst_path: pathlib.Path, content: bytes) -> None:
+    """Workaround for writing privileged files in destructive mode."""
+    try:
+        with tempfile.NamedTemporaryFile(delete=False) as f:
+            f.write(content)
+            f.flush()
+            f_name = f.name
+
+        try:
+            command = [
+                "sudo",
+                "install",
+                "--owner=root",
+                "--group=root",
+                "--mode=0644",
+                f_name,
+                str(dst_path),
+            ]
+            subprocess.run(command, check=True)
+        except subprocess.CalledProcessError:
+            raise RuntimeError(f"Failed to install repository config with: {command!r}")
+    finally:
+        os.unlink(f_name)
+
+
+def _get_host_arch() -> str:
+    return ProjectOptions().deb_arch
+
+
+class AptSourcesManager:
+    def __init__(
```

Comment:
It would be interesting to add a class docstring (and something for the public methods as well).